Fix torch.diagonal for torch.onnx.export when dim1<0 or dim2<0

veritas-Qiu opened the pull request:
in many cases, torch.diagonal will pass (dim1=-2, dim2=-1), onnx export will always fail in these cases
this pr try to fix the bug

veritas-Qiu committed:
fix a bug, onnx export will fail when dim1<0 or dim2<0 in torch.diagonal

thiagocrepaldi reviewed (commented):
LGTM. Can you add a unit test with negative dimensions to guard this from regression? Thank you for the contribution

veritas-Qiu committed:
Update test_operators.py

add unit test for diagonal

veritas-Qiu committed:
add expect result of test_diagonal

veritas-Qiu commented:
> LGTM. Can you add a unit test with negative dimensions to guard this from regression? Thank you for the contribution

Thanks.
Added.

thiagocrepaldi commented on torch/onnx/symbolic_opset13.py:
```diff
@@ -755,11 +755,14 @@ def repeat_interleave(
 @symbolic_helper.parse_args("v", "i", "i", "i")
 @_beartype.beartype
 def diagonal(g: jit_utils.GraphContext, self, offset, dim1, dim2):
+    rank = symbolic_helper._get_tensor_rank(self)
```
what if ` rank is None`? when the model is exported with ` torch.jit.script`, ` self.type()` could be `None`, forcing `_get_tensor_rank` to return None too

thiagocrepaldi commented on test/onnx/test_operators.py:
```diff
@@ -1289,6 +1289,10 @@ def forward(self, x):
             input_names=["x"],
             dynamic_axes={"x": {0: "dim_0"}},
         )
+    def test_diagonal(self):
+        x = torch.randn(2, 5, 4, 2)
+        self.assertONNX(lambda x: x.diagonal(offset=-1, dim1=-2, dim2=-1), x)
+
```
adding it to test/onnx/test_pytorch_onnx_onnxruntime.py would be better as a numerical validation would be done with the exported model

veritas-Qiu committed:
Update symbolic_opset13.py

opset9.size has already checked if dim < 0 , so it does not need to add the rank to the dims when calculating dim1_size and dim2_size, just check when rank is not None

veritas-Qiu committed:
Revert "Update test_operators.py"

This reverts commit 3b613e987b99fcccd304b661b7b938e44c1c336b.
move test to test_onnx_onnxruntime.py as review suggestions

veritas-Qiu committed:
Revert "add expect result of test_diagonal"

This reverts commit c6d84eb0754a8a964e28a64375741ce13be2ec2c.
move test to test_onnx_onnxruntime.py as review suggestions

veritas-Qiu committed:
move unit test to test_pytorch_onnx_onnxruntime.py

thiagocrepaldi commented on torch/onnx/symbolic_opset13.py:
```diff
@@ -771,9 +771,11 @@ def diagonal(g: jit_utils.GraphContext, self, offset, dim1, dim2):
-        axes.remove(dim1)
-        axes.remove(dim2)
-        self = g.op("Transpose", self, perm_i=axes + [dim1, dim2])
+        dim1_real = dim1 if dim1 >= 0 else dim1 + rank
```
Is it ok to have negative `dim1` and `dim2` on lines 759 and 762 above?

Would this work? This snippet re-computes `dim1` and `dim2` in the beginning if rank is available, otherwise, they remain as is. `dim1` and `dim2` would then be the same across the entire function

```python
@_onnx_symbolic("aten::diagonal")
@symbolic_helper.parse_args("v", "i", "i", "i")
@_beartype.beartype
def diagonal(g: jit_utils.GraphContext, self, offset, dim1, dim2):
    rank = symbolic_helper._get_tensor_rank(self)
    # Replace negative indexing when rank is known
    if rank is not None:
        dim1 = dim1 if dim1 >= 0 else dim1 + rank
        dim2 = dim2 if dim2 >= 0 else dim2 + rank

    dim1_size = opset9.size(
        g, self, dim=g.op("Constant", value_t=torch.LongTensor([dim1]))
    )
    dim2_size = opset9.size(
        g, self, dim=g.op("Constant", value_t=torch.LongTensor([dim2]))
    )
    # Create appropriate mask
    mask_shape = g.op("Concat", dim1_size, dim2_size, axis_i=0)
    mask = opset9.zeros(g, mask_shape, None, None, None)
    mask = g.op("EyeLike", mask, k_i=offset)
    # dim1 and dim2 appended as a dimension at the end of the shape
    rank = symbolic_helper._get_tensor_rank(self)
    if rank is not None:
        axes = list(range(rank))
        axes.remove(dim1)
        axes.remove(dim2)
        self = g.op("Transpose", self, perm_i=axes + [dim1, dim2])
    else:
        return symbolic_helper._unimplemented("diagonal", "unknown input rank")
```

veritas-Qiu committed:
Update symbolic_opset13.py

less code changes according to the review suggestions

veritas-Qiu committed:
Update symbolic_opset13.py

rank has been calculated 2 times

thiagocrepaldi reviewed (approved):
LGTM. Let's wait on the CI. Thanks

veritas-Qiu committed:
Update test_pytorch_onnx_onnxruntime.py

remove trailing spaces according to lint.

thiagocrepaldi commented:
@pytorchbot merge

thiagocrepaldi commented:
@pytorchbot merge -f "unrelated macos failure"
